[Markneoneo] iP

Markneoneo opened the pull request:
# Amadeus - CLI Task Manager

Amadeus is a command-line task manager designed to help you keep track of your daily tasks efficiently. Inspired by time-traveling themes, Amadeus organizes your todos, deadlines, and events in a structured format, allowing you to focus on productivity with minimal distractions.

## Features

- **ToDo Tasks**: Keep track of simple tasks.
- **Deadlines**: Assign due dates to tasks.
- **Events**: Schedule events with start and end times.
- **Mark/Unmark Tasks**: Track task completion.
- **Delete Tasks**: Remove completed or unnecessary tasks.
- **List Tasks**: View all current tasks in a categorized format.
- **Reset**: Clear all tasks instantly.
- **User-Friendly Prompts**: Provides clear and helpful feedback for user commands.
- **Steins;Gate Themed Exit Message**: A tribute to time-traveling storytelling.

Markneoneo committed:
Level 0. Rename, Greet, Exit.

Markneoneo committed:
Level 1. Echo

- Made Login and Logoff functions
- Included Echo function
- Exits with "bye"

Markneoneo committed:
Level 2. Add, List

- Added UserInput Class: To process and execute user input commands.
- Added TaskList Class: To store tasks and interact with the list.
- Added Task Classes: To establish properties of a task.
- Added the ability to store whatever text entered by the user and display them back to the user when requested.
- Added slow text printing for opening and closing.

Markneoneo committed:
Level 3. Mark as Done

- Added the ability to mark tasks as done.
- Added the ability to change the status back to not done.
- Added NumberConverter Class: provide ability to detect number words in the input
- Added proper documentation comments

Markneoneo committed:
Level 4. ToDos, Events, Deadlines

- Added Event Class: tasks that start at a specific date/time and ends at a specific date/time.
- Added Deadline Class: tasks that need to be done before a specific date/time.
- Added ToDo Class: tasks without any date/time attached to it.
- Created tasktypes Package: Task, Deadline, Todo, Event.
- Created userinterface Package: UserInput, TaskList, NumberConverter.
- Revamped Task List Printing and Documentation.

Markneoneo committed:
Automated Text UI Testing

- Set up the automated testing of text UIs using the input/output redirection technique.
- Updated runtest.bat to recognise emojis and skip Amadeus introduction during comparison.
- Created customised input tests to check all possible output variants.

Markneoneo committed:
Level 5. Handle Errors

- Added AmadeusException Class: Creates exceptions with custom error messages.
- Updated existing error messages to throw exceptions and catch them instead of printing.

Markneoneo committed:
Merge Level-5 increment

adoorknob commented on src/main/java/userinterface/NumberConverter.java:
```diff
@@ -0,0 +1,104 @@
+ * Ignores the word "and" for better parsing.
+ * Ensures incorrect inputs return -1 as an error indicator.
+ */
+public class NumberConverter
```
Impressive parser :)

adoorknob commented on src/main/java/Amadeus.java:
```diff
@@ -0,0 +1,126 @@
+     * @param text The text to be printed.
+     * @param duration The delay (in milliseconds) between each character.
+     */
+    public static void slowPrint(String text, int duration)
```
maybe start with a verb instead (as is the coding standard for methods)?

Markneoneo replied:
Gotcha I changed it to printSlow()

adoorknob commented on src/main/java/Amadeus.java:
```diff
@@ -0,0 +1,126 @@
+    public static void logIn()
+    {
+        System.out.println(DIVIDER);
+        slowPrint("User ID: Salieri\nLogging in...\nLaunching...\n", 50);
```
love the reference

adoorknob commented on src/main/java/userinterface/TaskList.java:
```diff
@@ -0,0 +1,191 @@
+                }
+                taskList.add(index, e); // Insert at the end of the Event section
+
+                System.out.println("✍️Understood! The following Event has been stored:\n╰┈➤ " + e);
```
maybe have a `return;` statement here? this way you can just have the exception throw outside to cover both cases and avoid the repeated code

Markneoneo replied:
It wouldn't work since 'return' is unnecessary as the last statement in the void function.
However, I do agree it looks repeated and messy, and decided to rewrite it to look cleaner!

adoorknob commented on src/main/java/userinterface/UserInput.java:
```diff
@@ -0,0 +1,137 @@
+    public final static String EVENT_COMMAND = "event";
+    public final static String TODO_COMMAND = "todo";
+    public static final String DIVIDER = "=".repeat(100);
+
```
i like the use of class level variables for consistency!

adoorknob commented on src/main/java/userinterface/UserInput.java:
```diff
@@ -0,0 +1,137 @@
+
+            // Adds a generic task using the full input as the task description.
+            default:
+                TaskList.storeTask(input);
```
should this also throw an error if the description is empty?

Markneoneo replied:
The empty description error is done in the UserInput, since it applies to all task commands!

adoorknob reviewed (commented):
lgtm just a few nits regarding method naming! very impressive and the references to steinsgate is quite fun haha! i've taken inspiration from your code also 🫡

Markneoneo committed:
A-Packages: Organize into Packages

- Reorganized the classes into suitable java packages.
- Added Peer Review Suggestions.

Markneoneo committed:
Merge A-Packages Increment

Markneoneo committed:
Level 6. Delete

- Added support for deleting tasks from the list.
- Updated automatic testing to test delete function.
- Added method for user to see all possible Amadeus commands.

Markneoneo committed:
Level 7. Save

- Save the tasks in the hard disk automatically whenever the task list changes. Load the data from the hard disk when the chatbot starts up.
- Added ability to fully "reset" saved list.
- Updated unit tests to include saving/loading.

swethacool commented on src/main/java/Amadeus.java:
```diff
@@ -42,7 +46,11 @@ public class Amadeus
+     * @param text String to be printed
+     * @param duration Time paused between each symbol
+     */
     public static void slowPrint(String text, int duration)
```
slowPrint() should be renamed to printSlow() to follow the verb-first naming convention.

swethacool commented on src/main/java/UserInput.java:
```diff
@@ -20,13 +29,35 @@ public static void getCommand()
+     * Scans the user input and executes their respective commands
+     * @param input Command input by user
+     */
     public static void scanInput(String input)
```
scanInput(String input) could be renamed to processInput() or executeCommand() to be more in line with the conventional naming practices for methods that handle user input or commands.

swethacool replied:
Good that it follows a verb-first naming convention!

swethacool commented on src/main/java/UserInput.java:
```diff
@@ -20,13 +29,35 @@ public static void getCommand()
+     * Scans the user input and executes their respective commands
+     * @param input Command input by user
+     */
     public static void scanInput(String input)
```
Suggestion to refactor scanInput() into smaller methods for better readability.

swethacool reviewed (commented):
Overall, the code is well-structured and functional. However minor refinements in naming will make it even better !

Markneoneo committed:
Marge Level-6 increment

Markneoneo committed:
Marge Level-7 increment

Markneoneo committed:
A-Jar: Create a JAR File

- Packaged the app as an executable JAR file so that it can be distributed easily.
- Updated commands to include 'reset'.
- String concatenation (+) in toString() creates multiple String objects, which is less efficient. Hence, I changed all to use String.format() for better readability and efficiency.

Markneoneo committed:
Merge A-Jar Increment

Markneoneo committed:
A-MoreOOP: Use More OOP

Refactored the code to extract out closely related code as classes.

- Database: Takes care of loading and saving tasks with the Memory.
- Speech: Prints all UI text interactions with the user.
- Parser: Deals with making sense of the user input.
- TaskList: Contains the task list and has operations to manipulate tasks in the list.
- AmadeusExceptions: Updated to use Static Factory Methods. This keeps all exception handling in a single class while allowing specific messages for different errors.
- Task: Completely removed functionality for Generic Misc. Tasks, and turned the class fully abstract. Now there are only 3 task types: Deadline, Event and Todo.
- Updated Test Cases to check new UI features.
- Updated all Documentation and comments.

Stretch Goal: Extracted multiple classes that inherit from an abstract Command class.

- Now every unique command has their own child class, and handles the entire logic and execution of the command within them.

Markneoneo committed:
Merge A-MoreOOP increment

Paulifyer commented on src/main/java/amadeus/brain/Database.java:
```diff
@@ -0,0 +1,140 @@
+        } catch (FileNotFoundException e) { // If the file is not found, start with an empty task list
+            System.out.println("⚠️ No existing task file found. Starting with an empty task list.");
+
+        } catch (Exception e) {
+            // Error Loading Task List from Memory.txt
+            throw AmadeusException.errorLoadingTask(e.getMessage());
+        }
```
Try not to simply catch all exceptions like this instead use the specific exception that may be thrown

Paulifyer commented on src/main/java/amadeus/perception/Parser.java:
```diff
@@ -0,0 +1,74 @@
+     * Constants representing valid user commands.
+     * These are used to map user input to the appropriate command object.
+     */
+    public final static String BYE_COMMAND = "bye";
+    public final static String COMMAND_LIST = "commands";
+    public final static String LIST_COMMAND = "list";
+    public final static String MARK_COMMAND = "mark";
+    public final static String UNMARK_COMMAND = "unmark";
+    public final static String DELETE_COMMAND = "delete";
+    public final static String RESET_COMMAND = "reset";
+    public final static String DEADLINE_COMMAND = "deadline";
+    public final static String EVENT_COMMAND = "event";
+    public final static String TODO_COMMAND = "todo";
```
you could consider using an enumerator here instead

Markneoneo committed:
Level 8. Dates and Times

- Taught the chatbot how to understand dates and times.
For example, if the command is deadline return book /by 2/12/2019 1800, the chatbot understands 2/12/2019 1800 as 2nd of December 2019, 6pm, instead of treating it as just a String.
- Stored deadline and event tasks dates as a java.time.LocalDate / java.time.LocalDateTime in my task objects.
Users can add tasks with a date only, or both a date and time.
- Strictly only accepted (dd-mm-yyyy) date format input and prints output in (dd MMM yyyy) format.
- Added Check Command: a command to print deadlines/events occurring on/before/after a specific date.

- Implemented flexible date and time formatting functionality.
Date can be input as: "1 1 2000" or "1/1/2000" or "1-1-2000"
Time can be input as: "1800" or "18:00" or "18.00" or "6pm" or "6:00PM" or "6.00pm"
- Updated exceptions to include invalid check commands, invalid time formats and invalid event times.
- Updated Unit tests and added testing documentation.

Markneoneo committed:
Level 9. Find

- Added Find Command: Given users a way to find a task by searching for a keyword in the task description.
- Updated Unit Tests.

Markneoneo committed:
Merge pull request #1 from Markneoneo/branch-Level-8

Level 8. Dates and Times

Markneoneo committed:
Merge branch 'master' into branch-Level-9

Markneoneo committed:
Merge pull request #2 from Markneoneo/branch-Level-9

Level 9. Find

Markneoneo committed:
A-JavaDoc: JavaDoc

- Added JavaDoc comments to the code.
Added header comments to all non-private classes/methods, and non-trivial private methods.

- Generated HTML Javadoc documentation for all Amadeus Java files into docs.

- Updated Test Cases and Markdown Documentation.

- Updated Speech to have more personality of Amadeus.

- Updated output messages to include ANSI colour coding and formatting.

- Updated Amadeus Logo (with colour now).

Markneoneo committed:
Merge pull request #3 from Markneoneo/branch-A-JavaDoc

A-JavaDoc: JavaDoc

Markneoneo committed:
A-UserGuide: User Guide

Updated README.md

Markneoneo committed:
Updated README.md

Markneoneo committed:
Update README.md

Markneoneo committed:
Update README.md

Markneoneo committed:
Added the Jekyll Hacker Theme to Github Pages

Markneoneo committed:
Merge pull request #4 from Markneoneo/hacker-theme

Added the Jekyll Hacker Theme to Github Pages

Markneoneo committed:
Update index.md

Markneoneo committed:
Update README.md

Markneoneo committed:
Locale Date Time BugFix

Markneoneo committed:
Added Checkstyle to automate formatting, linting and testing.

Markneoneo committed:
Reformatted Code.

Markneoneo committed:
Updated code to follow checkstyle.

Markneoneo committed:
Updated code to follow checkstyle.

Markneoneo committed:
Updated code to follow checkstyle.

Markneoneo committed:
Update README.md

Markneoneo committed:
Update README.md

Markneoneo committed:
Update index.md

Markneoneo committed:
Improved Amadeus Logo.

Markneoneo committed:
Merge branch 'master' of https://github.com/Markneoneo/ip

Markneoneo committed:
Update README.md

Markneoneo committed:
Update index.md

Markneoneo committed:
Update index.md

Markneoneo committed:
Update README.md

Markneoneo committed:
Update README.md

Markneoneo committed:
Update README.md

Markneoneo committed:
Update index.md

Markneoneo committed:
Updated Amadeus Logo

Markneoneo committed:
Update _config.yml

Markneoneo committed:
Update README.md
